### PR TITLE
Two Performance Improvement for Rebalance pipeline

### PR DESCRIPTION
**Pull request by dasahcc**
Two improvements have been implemented :
1. Avoid to read data from ZK for ExternalViewStage
2. Support selective update for IdealState

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java [hunk 1]
```diff
@@ -104,6 +108,52 @@ public void refresh(HelixDataAccessor accessor) {
...
+  /**
+   * Selective update Helix Cache by version
+   * @param accessor the HelixDataAccessor
+   * @param reloadKeys keys needs to be reload
+   * @param cachedKeys keys already exists in the cache
+   * @param cachedPropertyMap cached map of propertykey -> property object
+   * @param <T> the type of metadata
+   * @return
+   */
+  public static  <T extends HelixProperty> Map<PropertyKey, T> updateReloadProperties(
```
Comment (kishoreg) [inline helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java]: If this is a public static method, why is this in this class?

File: helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java [hunk 2]
```diff
@@ -104,6 +108,52 @@ public void refresh(HelixDataAccessor accessor) {
...
+      HelixDataAccessor accessor, List<PropertyKey> reloadKeys, List<PropertyKey> cachedKeys,
+      Map<PropertyKey, T> cachedPropertyMap) {
+    // All new entries from zk not cached locally yet should be read from ZK.
+    Map<PropertyKey, T> refreshedPropertyMap = Maps.newHashMap();
+    List<HelixProperty.Stat> stats = accessor.getPropertyStats(cachedKeys);
+    for (int i = 0; i < cachedKeys.size(); i++) {
```
Comment (kishoreg) [inline helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java]: what will the stat be if the znode is deleted

File: helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java [hunk 3]
```diff
@@ -104,6 +108,52 @@ public void refresh(HelixDataAccessor accessor) {
...
+          reloadKeys.add(key);
+        }
+      } else {
+        LOG.warn("stat is null for key: " + key);
+        reloadKeys.add(key);
+      }
+    }
+
+    List<T> reloadedProperty = accessor.getProperty(reloadKeys, true);
+    Iterator<PropertyKey> csKeyIter = reloadKeys.iterator();
```
Comment (kishoreg) [inline helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java]: is this method specific to current state? why are the variables named cs?

File: helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java [hunk 4]
```diff
@@ -104,6 +108,52 @@ public void refresh(HelixDataAccessor accessor) {
...
+    for (T property : reloadedProperty) {
+      PropertyKey key = csKeyIter.next();
+      if (property != null) {
+        refreshedPropertyMap.put(key, property);
+      } else {
+        LOG.warn("CurrentState null for key: " + key);
```
Comment (kishoreg) [inline helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java]: this is not related to currentstate right

File: helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java [hunk 5]
```diff
@@ -104,6 +108,52 @@ public void refresh(HelixDataAccessor accessor) {
...
+   * Selective update Helix Cache by version
+   * @param accessor the HelixDataAccessor
+   * @param reloadKeys keys needs to be reload
+   * @param cachedKeys keys already exists in the cache
+   * @param cachedPropertyMap cached map of propertykey -> property object
+   * @param <T> the type of metadata
+   * @return
+   */
+  public static  <T extends HelixProperty> Map<PropertyKey, T> updateReloadProperties(
+      HelixDataAccessor accessor, List<PropertyKey> reloadKeys, List<PropertyKey> cachedKeys,
```
Comment (kishoreg) [inline helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java]: how will cachedKeys be different from cachedPropertyMap.keySet?

File: helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java (hunk 1 above)
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java]: Since this method will be used by different cache class, I think it is better to put this method the basic class of them.

File: helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java (hunk 2 above)
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java]: deleted the node has been removed for previous preprocess. The cached key will only retain the keys from reload keys from ZK.

File: helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java (hunk 3 above)
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java]: Will change the name.

File: helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java (hunk 4 above)
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java]: Good catch. Will make correct logs.

File: helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java (hunk 5 above)
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java]: cachedkeys only contains the existing keys. Deleted node will be removed from it.
